test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresUnreferencedCodeCapability.cs

```diff
@@ -1016,15 +1143,197 @@ static void TestSuppressionsOnClass ()
+					private static int PrivateStaticField;
+				}
+
+				[RequiresUnreferencedCode ("--WithRUCOnlyInstanceFields--")]
+				class WithRUCOnlyInstanceFields
+				{
+					public int InstnaceField;
+				}
+
+				[ExpectedWarning ("IL2109", "DAMAccessOnField/DerivedWithoutRUC", "DAMAccessOnField.WithRUC", GlobalAnalysisOnly = true)]
+				class DerivedWithoutRUC : WithRUC
+				{
+					public static int DerivedStaticField;
+				}
+
+				[RequiresUnreferencedCode ("--DerivedWithRUC--")]
+				class DerivedWithRUC : WithRUC
+				{
+					public static int DerivedStaticField;
+				}
+
+				[ExpectedWarning ("IL2026", "WithRUC::StaticField", GlobalAnalysisOnly = true)]
+				[ExpectedWarning ("IL2026", "WithRUC::PrivateStaticField", GlobalAnalysisOnly = true)]
+				[ExpectedWarning ("IL2026", "DerivedWithRUC::DerivedStaticField", GlobalAnalysisOnly = true)]
+				public static void Test ()
+				{
+					typeof (WithRUC).RequiresPublicFields ();
+					typeof (WithRUC).RequiresNonPublicFields ();
+					typeof (WithRUCOnlyInstanceFields).RequiresPublicFields ();
+					typeof (DerivedWithoutRUC).RequiresPublicFields ();
+					typeof (DerivedWithRUC).RequiresPublicFields ();
+				}
+			}
+
+			[RequiresUnreferencedCode ("Field use is dangerous")]
+			public class ClassWithRUCFieldUsedViaReflection
+			{
+				public static int publicField;
+			}
+
+			[ExpectedWarning ("IL2026", "FieldUsedViaReflection::publicField", GlobalAnalysisOnly = true)]
+			static void KeepFieldViaReflection ()
+			{
+				var field = typeof (ClassWithRUCFieldUsedViaReflection).GetField ("publicField");
+				field.GetValue (null);
+			}
+
+			[ExpectedWarning ("IL2026", "FieldUsedViaReflection::publicField", GlobalAnalysisOnly = true)]
+			[DynamicDependency ("publicField", typeof (ClassWithRUCFieldUsedViaReflection))]
+			static void KeepFieldViaDynamicDependency () { }
+
+			[RequiresUnreferencedCode ("The attribute is dangerous")]
+			public class AttributeWithRUC : Attribute
+			{
+				public static int field;
+
+				// `field` cannot be used as named attribute argument because is static, and if accessed via
+				// a property the property will be the one generating the warning, but then the warning will 
+				// be suppresed by the RequiresUnreferencedCode on the declaring type
+				public int PropertyOnAttribute {
+					get { return field; }
+					set { field = value; }
+				}
+			}
+
+			[AttributeWithRUC (PropertyOnAttribute = 42)]
+			[ExpectedWarning ("IL2026", "AttributeWithRUC.AttributeWithRUC()", GlobalAnalysisOnly = true)]
+			static void KeepFieldOnAttribute () { }
+
+			[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicFields | DynamicallyAccessedMemberTypes.NonPublicFields)]
+			[RequiresUnreferencedCode ("This class is dangerous")]
+			class ClassImplementingAnnotatedInterface
+			{
+				[ExpectedWarning ("IL2112", "ClassImplementingAnnotatedInterface::publicField", GlobalAnalysisOnly = true)]
+				public static int publicField;
```

Comment:
I wasn't fully sure if the field is the best place to warn about IL2112

Reply:
This is at least consistent with what we do for `DAM`-annotated fields, so I think it's the right place for the warning. Though I agree this could be confusing.

Reply:
I think this is fine - it's the same for methods. The IL2112 and similar warnings are inherently confusing to some degree, hopefully they should be super rare.